pkg/util/podstore/podstore.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package podstore
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
+)
+
+const (
+	deleteQueueName = "pod_delete"
+	podIPIndex      = "PodIP"
+	delayTime       = time.Minute * 5
+)
+
+type PodStore struct {
+	pods         cache.Indexer
+	podsToDelete workqueue.DelayingInterface
+	// Mapping pod.uuid to PodLocalTimestamps
+	timestampMap map[types.UID]*podLocalTimestamps
+	clock        clock.Clock
+	mutex        sync.Mutex
+}
+
+type podLocalTimestamps struct {
+	// The local clock when the Pod is added
+	CreationTimestamp time.Time
+	// The local clock when the Pod is actually deleted from kube-apiserver. Nil if it's not deleted yet.
+	DeletionTimestamp *time.Time
+}
+
+// Interface is a podStore interface to create local podStore for Flow Exporter and Flow Aggregator.
+type Interface interface {
+	GetPodByIPAndTime(ip string, startTime time.Time) (*corev1.Pod, bool)
+	Run(stopCh <-chan struct{})
+}
+
+// NewPodStoreWithClock creates a Pod Store with a custom clock,
+// which is useful when writing robust unit tests.
+func NewPodStoreWithClock(podInformer cache.SharedIndexInformer, clock clock.WithTicker) *PodStore {
+	s := &PodStore{
+		pods:         cache.NewIndexer(podKeyFunc, cache.Indexers{podIPIndex: podIPIndexFunc}),
+		podsToDelete: workqueue.NewDelayingQueueWithCustomClock(clock, deleteQueueName),
+		clock:        clock,
+		timestampMap: map[types.UID]*podLocalTimestamps{},
+	}
+	podInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    s.onPodCreate,
+			UpdateFunc: s.onPodUpdate,
+			DeleteFunc: s.onPodDelete,
+		})
+	return s
+}
+
+func NewPodStore(podInformer cache.SharedIndexInformer) *PodStore {
+	return NewPodStoreWithClock(podInformer, clock.RealClock{})
+}
+
+func (s *PodStore) onPodUpdate(oldObj interface{}, newObj interface{}) {
+	newPod, ok := newObj.(*corev1.Pod)
+	if !ok {
+		klog.ErrorS(nil, "Received unexpected object", "newObj", newObj)
+		return
+	}
+	s.pods.Update(newPod)
+	klog.V(4).InfoS("Process Pod Update Event", "Pod", klog.KObj(newPod))
```

Comment:
should be "Processed" and not "Process" for all these log messages, since you log them at the end of processing

Reply:
Fixed.